nomos-services/data-availability/network/src/opinion_aggregator.rs

```diff
@@ -0,0 +1,238 @@
+        // Generate opinions before clearing if we have both memberships
+        let opinions = if self.current_membership.is_some() && self.previous_membership.is_some() {
+            self.generate_opinions()
+        } else {
+            None
+        };
+
+        self.previous_membership = self.current_membership.take();
+
+        self.positive_opinions.clear();
+        self.negative_opinions.clear();
+        self.blacklist.clear();
+        self.old_positive_opinions.clear();
+        self.old_negative_opinions.clear();
+        self.old_blacklist.clear();
+
+        // Pre-populate opinion maps with zeros to the right vector size
+        for peer_id in new_membership.members() {
+            self.positive_opinions.insert(peer_id, 0);
+            self.negative_opinions.insert(peer_id, 0);
+        }
+
+        if let Some(ref membership) = self.previous_membership {
+            for peer_id in membership.members() {
+                self.old_positive_opinions.insert(peer_id, 0);
+                self.old_negative_opinions.insert(peer_id, 0);
+            }
+        }
+
+        self.current_membership = Some(new_membership);
+
+        // Return opinions to be sent to SDP
+        // SDP will convert peerID -> ProviderID and sort in lexicographical order
+        opinions
+    }
+
+    pub fn generate_opinions(&self) -> Option<Opinions> {
+        let current = self.current_membership.as_ref()?;
+        let previous = self.previous_membership.as_ref()?;
+
+        let new_opinions = self.calculate_opinions_map(
+            current.members().into_iter(),
+            &self.positive_opinions,
+            &self.negative_opinions,
+            true,
+        );
+
+        let old_opinions = self.calculate_opinions_map(
+            previous.members().into_iter(),
+            &self.old_positive_opinions,
+            &self.old_negative_opinions,
+            previous
+                .members()
+                .into_iter()
+                .any(|id| id == self.local_peer_id),
+        );
+
+        Some(Opinions {
+            session_id: current.session_id(),
+            new_opinions,
+            old_opinions,
+        })
+    }
+
+    fn calculate_opinions_map(
+        &self,
+        peers: impl Iterator<Item = PeerId>,
+        positive: &HashMap<PeerId, u32>,
+        negative: &HashMap<PeerId, u32>,
+        include_self: bool,
+    ) -> HashMap<PeerId, bool> {
+        peers
+            .map(|peer_id| {
+                let opinion = if include_self && peer_id == self.local_peer_id {
+                    true
+                } else {
+                    let pos = positive.get(&peer_id).copied().unwrap_or(0);
+                    let neg = negative.get(&peer_id).copied().unwrap_or(0);
+                    pos > neg * OPINION_THRESHOLD
+                };
+                (peer_id, opinion)
+            })
+            .collect()
+    }
+}
+
+impl Opinions {
```

Comment:
Nit: lets leave it as is, but the Activity message will be formed by the SDP service, the encoding will be defined in nomos-core, this will be changed when working on SDP service.

Reply:
I agree, this was a helper fn with the assumption that sdp will provide mappins for PeerID -> ProviderID. Should be adapter when implementing.

Reply:
@danielSanchezQ suggested this helper functions as his idea was for sdp to figure out the mappings based on locators. If this is not possible then we can work it out in da service by either saving the appropriate mappings to storage whil processing updates, or by saving the actual assignations per ProviderID and creating the PeerID from ProviderID only when it is needed by behaviors

Reply:
> @danielSanchezQ what do you think if DA network will provide SDP service opinions already sortable (or sorted) by the provider ids.
 
 Yes that was the initial idea, but we do not have the provider ids in the network level from my last convo with @pradovic.
 If we can have it, then we should use the bitarrays for opinions and should be sorted.

Reply:
> Yes that was the initial idea, but we do not have the provider ids

We actually receive ProviderId from membership update stream, just convert it to PeerId inside DA network service, instead of converting it to PeerId, we'll create a mapping in DA network service.

Reply:
@danielSanchezQ I proposed to have mappings in the da service, but I understood that you were against it 🙈 We would need to keep track of the mappings. Then we could provide them sorted

Reply:
@danielSanchezQ good to merge then? Or should we first open a da related pr to handle the mappings?

Reply:
I agree.